src/sapling/sapling_operation.cpp

```diff
@@ -316,26 +316,42 @@ OperationResult SaplingOperation::loadUnspentNotes(TxValues& txValues,
                                                    uint256& ovk)
 {
     // if we already have selected the notes, let's directly se them.
-    if (coinControl && coinControl->HasSelected()) {
-        // todo: add notes selection.
-    }
+    bool hasCoinControl = coinControl && coinControl->HasSelected();
+    if (hasCoinControl) {
+        std::vector<OutPointWrapper> vCoins;
+        coinControl->ListSelected(vCoins);
 
-    std::vector<SaplingNoteEntry> saplingEntries;
-    pwalletMain->GetSaplingScriptPubKeyMan()->GetFilteredNotes(saplingEntries, fromAddress.fromSapAddr, mindepth);
-
-    for (const auto& entry : saplingEntries) {
-        shieldedInputs.emplace_back(entry);
-        std::string data(entry.memo.begin(), entry.memo.end());
-        LogPrint(BCLog::SAPLING,"%s: found unspent Sapling note (txid=%s, vShieldedSpend=%d, amount=%s, memo=%s)\n",
-                 __func__ ,
-                 entry.op.hash.ToString().substr(0, 10),
-                 entry.op.n,
-                 FormatMoney(entry.note.value()),
-                 HexStr(data).substr(0, 10));
-    }
+        // Converting outpoint wrapper to sapling outpoints
+        std::vector<SaplingOutPoint> vSaplingOutpoints;
+        for (const auto& outpoint : vCoins) {
+            vSaplingOutpoints.emplace_back(outpoint.outPoint.hash, outpoint.outPoint.n);
+        }
 
-    if (shieldedInputs.empty()) {
-        return errorOut("Insufficient funds, no available notes to spend");
+        pwalletMain->GetSaplingScriptPubKeyMan()->GetNotes(vSaplingOutpoints, shieldedInputs);
+
+        if (shieldedInputs.empty()) {
+            return errorOut("Insufficient funds, no available notes to spend");
+        }
+    } else {
+        // If we don't have coinControl then let's find the notes
+        std::vector<SaplingNoteEntry> saplingEntries;
+        pwalletMain->GetSaplingScriptPubKeyMan()->GetFilteredNotes(saplingEntries, fromAddress.fromSapAddr, mindepth);
+
+        for (const auto& entry : saplingEntries) {
+            shieldedInputs.emplace_back(entry);
+            std::string data(entry.memo.begin(), entry.memo.end());
+            LogPrint(BCLog::SAPLING,
+                     "%s: found unspent Sapling note (txid=%s, vShieldedSpend=%d, amount=%s, memo=%s)\n",
+                     __func__,
+                     entry.op.hash.ToString().substr(0, 10),
+                     entry.op.n,
+                     FormatMoney(entry.note.value()),
+                     HexStr(data).substr(0, 10));
+        }
+
+        if (shieldedInputs.empty()) {
+            return errorOut("Insufficient funds, no available notes to spend");
```

Comment:
nit: unify this check and the one at lines 332-333, by putting it outside  of `if (hasCoinControl) {} else {}`.